### PR TITLE
[LOGS]: Added the possibility to dualship logs

### PR DESCRIPTION
**Pull request by ajacquemot**
### What does this PR do?

Changed the sender to always send logs to a main endpoint, keep retrying until logs are effectively sent and added additional endpoints to send logs with a best effort strategy, try once and forget.

### Motivation

Internal use only.

### Additional Notes

- [x] Write unit-tests
- [x] Add comments

### CODE REVIEW

File: pkg/logs/config/endpoints.go
Comment (mnshdw) [inline pkg/logs/config/endpoints.go]: I wish we could avoid this, I think we can live without it now that we have index filters.

**Review by mnshdw — DISMISSED** [review]
Nice change. I think we could strengthen the naming by having Clients +  Client or Destinations + Destination but up to you. The only complaint that I have is `logset`, can we avoid using it in new code ?

File: pkg/logs/config/endpoints.go
Comment (tmichelet) [reply pkg/logs/config/endpoints.go]: +1 but it'll be easy to drop, and I don't think today we're fully ready

File: pkg/logs/config/endpoints.go
Comment (tmichelet) [reply pkg/logs/config/endpoints.go]: I'll let @ajacquemot answer but things to keep in mind are that we don't want to fully support this (today) so we try to keep it simple to fit our use case

File: pkg/logs/sender/destinations.go
Comment (ajacquemot) [reply pkg/logs/sender/destinations.go]: https://en.wiktionary.org/wiki/additionals :)

File: pkg/logs/config/endpoints.go
Comment (ajacquemot) [reply pkg/logs/config/endpoints.go]: The idea here is to send logs to multiple datadog accounts, so we should transport them the same way.
For the proxy address, the main use case is an organisation that wants to have a look at what information is leaking out so they just need one stream of logs I think.

File: pkg/logs/pipeline/pipeline.go
Comment (ajacquemot) [reply pkg/logs/pipeline/pipeline.go]: The goal is to send logs to Datadog for every endpoint, this looks like another feature.

File: pkg/logs/sender/client.go
Comment (ajacquemot) [reply pkg/logs/sender/client.go]: good catch

File: pkg/logs/config/endpoints.go
Comment (ajacquemot) [reply pkg/logs/config/endpoints.go]: adding a comment though

File: pkg/logs/pipeline/pipeline.go
Comment (tmichelet) [reply pkg/logs/pipeline/pipeline.go]: good point for a future proof version, for today let's not be too sophisticated :)

File: pkg/logs/sender/sender.go
Comment (tmichelet) [inline pkg/logs/sender/sender.go]: Could you explain what we're addressing with this `FramingError`? why do we want to fix it now?

**Comment by tmichelet** [comment]
This looks great, thanks @ajacquemot ! One question, what are we catching with the new exception handling logic? And why does it need to happen in this PR?

File: pkg/logs/sender/destination.go
Comment (iksaif) [inline pkg/logs/sender/destination.go]: Here we have delimiter and frame. maybe try to use the same word

File: pkg/logs/config/endpoints.go
Comment (iksaif) [inline pkg/logs/config/endpoints.go]: I think it would be easier to put SSL and Proxy here.

File: pkg/logs/pipeline/pipeline.go
Comment (iksaif) [inline pkg/logs/pipeline/pipeline.go]: looks like the `NewDestination()` from and endpoint could be factorized (it looks a lot like the one above)

File: pkg/logs/sender/connection_manager.go
Comment (iksaif) [inline pkg/logs/sender/connection_manager.go]: Looks like a `ConnectionManager` could contain or extend and `EndPoint`

**Review by iksaif — CHANGES_REQUESTED** [review]

File: pkg/logs/sender/connection_manager.go
Comment (ajacquemot) [reply pkg/logs/sender/connection_manager.go]: 👍

File: pkg/logs/sender/destination.go
Comment (ajacquemot) [reply pkg/logs/sender/destination.go]: Agreed we could merge both the prefixer and the delimeter. I'll open a PR right away after merging this one

File: pkg/logs/config/config.go
Comment (iksaif) [inline pkg/logs/config/config.go]: I would set UseProto here too, it's very similar to SSL.

File: pkg/logs/sender/destination.go
Comment (iksaif) [reply pkg/logs/sender/destination.go]: 👍

**Review by iksaif — CHANGES_REQUESTED** [review]
Minor comments + we need a unit test (I would feel safer if we have one that really opens a connection like : https://github.com/DataDog/datadog-agent/pull/2373/files#diff-05fd3fce70913c5d7acb557b89527d12 )

File: pkg/logs/config/config.go
Comment (ajacquemot) [reply pkg/logs/config/config.go]: 👍

File: pkg/logs/sender/prefixer_test.go
Comment (iksaif) [inline pkg/logs/sender/prefixer_test.go]: typo ?

**Review by iksaif — CHANGES_REQUESTED** [review]

**Review by iksaif — DISMISSED** [review]

**Comment by iksaif** [comment]
Approved, we will improve this and add more tests once merged with my PR. Wait for confirmation from somebody else.

File: pkg/logs/sender/destinations.go
Comment (achntrl) [reply pkg/logs/sender/destinations.go]: I meant `Additonals` -> `AdditIonals`, there is a missing `i`

**Review by mnshdw — DISMISSED** [review]

**Review by mnshdw — APPROVED** [review]

**Comment by remh** [comment]
That's great! Thank you!

**Pull request merged**
